tools/perf/builtin-c2c.c

```diff
@@ -1773,6 +2052,199 @@ static struct c2c_dimension dim_mean_lcl_peer = {
+static uint64_t get_total_cycles_all_symbols(void)
+{
+	struct rb_node *nd;
+	uint64_t total_cycles = 0;
+
+	nd = rb_first_cached(&c2c.symbol_hists.hists.entries);
+	while (nd) {
+		struct hist_entry *he = rb_entry(nd, struct hist_entry, rb_node);
+		struct c2c_hist_entry *c2c_he = container_of(he, struct c2c_hist_entry, he);
+
+		total_cycles += calculate_symbol_total_cycles(c2c_he);
+		nd = rb_next(nd);
+	}
+	return total_cycles;
+}
+
+static int
+cycles_percent_entry(struct perf_hpp_fmt *fmt, struct perf_hpp *hpp,
+		     struct hist_entry *he)
+{
+	struct c2c_hist_entry *c2c_he;
+	int width = c2c_width(fmt, hpp, he->hists);
+	uint64_t symbol_cycles;
+	uint64_t total_cycles;
+	double percent;
+
+	/* Hide Cycles Percent for child symbols and cachelines */
+	if (he->parent_he) {
+		return scnprintf(hpp->buf, hpp->size, "%*s", width, "");
+	}
+
+    c2c_he = container_of(he, struct c2c_hist_entry, he);
+    symbol_cycles = calculate_symbol_total_cycles(c2c_he);
+
+    /* For child symbols, scope denominator to related cachelines (siblings under parent) */
+    if (he->parent_he) {
+        struct c2c_hist_entry *parent_c2c_he;
+        struct related_symbol *rel_sym;
+        uint64_t denom_cycles = 0;
+
+        parent_c2c_he = container_of(he->parent_he, struct c2c_hist_entry, he);
+        list_for_each_entry(rel_sym, &parent_c2c_he->related_symbols, list) {
+            uint64_t rel_cycles_rmt = avg_stats(&rel_sym->cstats.rmt_hitm) * rel_sym->stats.rmt_hitm;
+            uint64_t rel_cycles_lcl = avg_stats(&rel_sym->cstats.lcl_hitm) * rel_sym->stats.lcl_hitm;
+            uint64_t rel_other_load = rel_sym->stats.load - rel_sym->stats.rmt_hitm - rel_sym->stats.lcl_hitm;
+            uint64_t rel_cycles_load = avg_stats(&rel_sym->cstats.load) * rel_other_load;
+            denom_cycles += rel_cycles_rmt + rel_cycles_lcl + rel_cycles_load;
+        }
+
+        percent = denom_cycles ? (double)symbol_cycles / denom_cycles * 100.0 : 0.0;
+    } else {
+        total_cycles = get_total_cycles_all_symbols();
+        percent = total_cycles > 0 ? (double)symbol_cycles / total_cycles * 100.0 : 0.0;
+    }
+
+    if (he->parent_he) {
+        /* Indent child percentages by 4 spaces but preserve column width */
+        char out[32];
+        scnprintf(out, sizeof(out), "    %.2f%%", percent);
+        return scnprintf(hpp->buf, hpp->size, "%-*s", width, out);
+    }
+
+	return scnprintf(hpp->buf, hpp->size, "%*.2f%%", width-1, percent);
+}
+
+/* Comparison function for cycles percentage sorting */
+static int64_t
+cycles_percent_cmp(struct perf_hpp_fmt *fmt __maybe_unused,
+		   struct hist_entry *left, struct hist_entry *right)
+{
+	struct c2c_hist_entry *c2c_left = container_of(left, struct c2c_hist_entry, he);
+	struct c2c_hist_entry *c2c_right = container_of(right, struct c2c_hist_entry, he);
+	uint64_t cycles_left, cycles_right;
+
+	cycles_left = calculate_symbol_total_cycles(c2c_left);
+	cycles_right = calculate_symbol_total_cycles(c2c_right);
```

Comment:
to speedup, can we store total cycles as an attribute to hist_entry instread of doing calculation each time.